qcodes/instrument/parameter.py

```diff
@@ -947,6 +947,90 @@ def sweep(self, start, stop, step=None, num=None):
                                 step=step, num=num)
 
 
+class ParameterWithSetpoints(Parameter):
+    """
+    A parameter that has associated setpoints. The setpoints is nothing
+    more than a list of other parameters that describe the values, names
+    and units of the setpoint axis for this parameter.
+
+    In most cases this will probably be a parameter that returns an array.
+    It is expected that the setpoint arrays are 1D arrays such that the
+    combined shape of the parameter. E.G if parameter is of shape (m,n)
+    self.setpoints is a list of parameters of shape (m,) and (n,)
+
+    In all other ways this is identical to  :class:`Parameter` See the
+    documentation of :class:`Parameter` for more details.
+    """
+    def __init__(self, *args,
+                 setpoints: Optional[Sequence[_BaseParameter]]=None,
+                 **kwargs) -> None:
+        if setpoints is None:
+            self._setpoints = []
+        else:
+            self.setpoints = setpoints
+        super().__init__(*args, **kwargs)
+
+    @property
+    def setpoints(self):
+        return self._setpoints
+
+    @setpoints.setter
+    def setpoints(self, setpoints: Sequence[_BaseParameter]):
+        for setpointarray in setpoints:
+            if not isinstance(setpointarray, _BaseParameter):
+                raise RuntimeError
+        self._setpoints = setpoints
+
+    def validate_consistent_shape(self) -> bool:
+        """
+        Verifies that the shape of the Array Validator of the parameter
+        is consistent with the Validator of the Setpoints. This requires that
+        both the setpoints and the actual parameters have validators
+        of type Arrays with a defined shape.
+        """
+        output_shape = self.vals.shape
+
+        if not isinstance(self.vals, Arrays):
+            raise RuntimeError("Can only validate shapes for Array "
+                               "Parameters")
+
+        setpoints_shape_list = []
+        for sp in self.setpoints:
+            if not isinstance(self.vals, Arrays):
+                raise RuntimeError("Can only validate shapes for Array "
```

Comment:
I don't understand this. Why are you doing this is a loop? Is `self.vals` somehow changing in the loop through some side effect (I hope not!)

Reply:
It's obviously a typo it should be the validator of the setpoints